maven-embedder/src/main/java/org/apache/maven/cli/RemoteRepositoryConnectionVerifier.java

```diff
@@ -0,0 +1,130 @@
+        this.transporterProvider = container.lookup(TransporterProvider.class);
+    }
+
+    private boolean isCentralOrMirrorOfCentral(final RemoteRepository remoteRepository) {
+        return "central".equals(remoteRepository.getId())
+                || remoteRepository.getMirroredRepositories().stream()
+                        .map(RemoteRepository::getId)
+                        .anyMatch("central"::equals);
+    }
+
+    public Optional<String> verifyConnectionToRemoteRepository(
+            final RepositorySystemSession session, final ArtifactRepository artifactRepository) {
+        final RemoteRepository repository = RepositoryUtils.toRepo(artifactRepository);
+
+        final String artifactPath;
+
+        if (isCentralOrMirrorOfCentral(repository)) {
+            // We can be sure the Apache Maven artifact should be resolvable.
+            artifactPath = artifactRepository.getLayout().pathOf(RepositoryUtils.toArtifact(APACHE_MAVEN_ARTIFACT));
+        } else {
+            // We cannot be sure about any artifact that lives here.
+            artifactPath = "";
+        }
+
+        try {
+            final Transporter transporter = transporterProvider.newTransporter(session, repository);
+            final Optional<String> maybeIssue = verifyConnectionUsingTransport(transporter, repository, artifactPath);
+
+            if (!maybeIssue.isPresent()) {
+                logger.info("Connection check for {} [{}] completed", repository.getId(), repository.getUrl());
+            }
+
+            return maybeIssue;
+        } catch (final NoTransporterException nte) {
+            final String message = String.format(
+                    "There is no compatible transport for remote repository %s with location %s",
+                    repository.getId(), repository.getUrl());
+            return Optional.of(message);
+        }
+    }
+
+    private Optional<String> verifyConnectionUsingTransport(
+            final Transporter transporter, final RemoteRepository remoteRepository, final String artifactPath) {
+        try {
+            final GetTask task = new GetTask(URI.create(artifactPath));
+            transporter.get(task);
+            return Optional.empty();
+        } catch (final Exception e) {
+            return classifyException(remoteRepository, e);
+        }
+    }
+
+    private Optional<String> classifyException(final RemoteRepository remoteRepository, final Exception e) {
+        final String message = e.getMessage();
+        final String repositoryId = remoteRepository.getId();
+        final String repositoryUrl = remoteRepository.getUrl();
+        final String repository = String.format("%s [%s]", repositoryId, repositoryUrl);
+
+        final boolean notFound = StringUtils.contains(message, "status code: 404");
+        final boolean unauthorized = StringUtils.contains(message, "status code: 401");
+        final boolean forbidden = StringUtils.contains(message, "status code: 403");
```

Comment:
Uh, that really hurts. No better way? You assume that transport is always HTTP? Either in Wagon or Resolver there is a classify method, maybe that can help...

Reply:
We now use [`transport#classify()`](https://maven.apache.org/resolver/apidocs/org/eclipse/aether/spi/connector/transport/Transporter.html#classify(java.lang.Throwable)). It does not provide a lot of detail, but it's transport-agnostic AFAICS, which is definitely a plus.